lib/puppet/pops/loaders.rb

```diff
@@ -51,12 +51,13 @@ def initialize(environment, for_agent)
     #    TODO: loaders need to work when also running in an agent doing catalog application. There is no
     #    concept of environment the same way as when running as a master (except when doing apply).
     #    The creation mechanisms should probably differ between the two.
-    @private_environment_loader = if for_agent
-      @puppet_cache_loader = create_puppet_cache_loader()
-      create_environment_loader(environment, @puppet_cache_loader)
-    else
-      create_environment_loader(environment, @puppet_system_loader)
-    end
+    @private_environment_loader =
+      if for_agent
+        @puppet_cache_loader = create_puppet_cache_loader
+        create_environment_loader(environment, @puppet_cache_loader, load_from_pcore)
```

Comment:
Before and after this change, the `puppet agent` the `Runtime3xTypeLoader` creates a pcore resource type loader, which seems unnecessary and could lead to odd behavior when running the agent on the puppetserver and pcore resource types exist. Also why does DeferredResolver create a new set of loaders instead of using the existing loaders on the environment?

Reply:
The script compiler accesses the `catalog.environment_instance` which is a different remote environment instance than the one the configurer uses, so both end up creating loaders. The slowdown isn't too noticeable though if a trivial catalog...

Reply:
It seems unnecessary to store both the `:current_environment` and `:loaders` in puppet's context as it sometimes leads to surprising behavior when the `:loaders` are bound to a different environment object than the one pushed on the context. I changed `lib/puppet/pops/{adapters,loaders}.rb` to lookup the loaders from the `:current_environment` and removed the calls that pushed `:loaders`. Agent and apply worked with deferred and sensitive data. So maybe something we can clean up in puppet 7